annotation.go

```diff
@@ -0,0 +1,213 @@
+	PanelID     int64    `json:"panelId"`
+	UserID      int64    `json:"userId,omitempty"`
+	UserName    string   `json:"userName,omitempty"`
+	NewState    string   `json:"newState,omitempty"`
+	PrevState   string   `json:"prevState,omitempty"`
+	Time        int64    `json:"time"`
+	TimeEnd     int64    `json:"timeEnd,omitempty"`
+	Text        string   `json:"text"`
+	Metric      string   `json:"metric,omitempty"`
+	RegionID    int64    `json:"regionId,omitempty"`
+	Type        string   `json:"type,omitempty"`
+	Tags        []string `json:"tags,omitempty"`
+	IsRegion    bool     `json:"isRegion,omitempty"`
+}
+
+// GraphiteAnnotation represents a Grafana API annotation in Graphite format
+type GraphiteAnnotation struct {
+	What string   `json:"what"`
+	When int64    `json:"when"`
+	Data string   `json:"data"`
+	Tags []string `json:"tags,omitempty"`
+}
+
+// Annotations fetches the annotations queried with the params it's passed
+func (c *Client) Annotations(params url.Values) ([]Annotation, error) {
+	req, err := c.newRequest("GET", "/api/annotation", params, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, errors.New(resp.Status)
+	}
+
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	result := []Annotation{}
+	err = json.Unmarshal(data, &result)
+	return result, err
+}
+
+// NewAnnotation creates a new annotation with the Annotation it is passed
+func (c *Client) NewAnnotation(a *Annotation) (int64, error) {
+	data, err := json.Marshal(a)
+	if err != nil {
+		return 0, err
+	}
+	req, err := c.newRequest("POST", "/api/annotations", nil, bytes.NewBuffer(data))
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return 0, err
+	}
+	if resp.StatusCode != 200 {
+		return 0, errors.New(resp.Status)
+	}
+
+	data, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, err
+	}
+
+	result := struct {
+		ID int64 `json:"id"`
+	}{}
+	err = json.Unmarshal(data, &result)
+	return result.ID, err
+}
+
+// NewGraphiteAnnotation creates a new annotation with the GraphiteAnnotation it is passed
+func (c *Client) NewGraphiteAnnotation(gfa *GraphiteAnnotation) (int64, error) {
+	data, err := json.Marshal(gfa)
+	if err != nil {
+		return 0, err
+	}
+	req, err := c.newRequest("POST", "/api/annotations/graphite", nil, bytes.NewBuffer(data))
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return 0, err
+	}
+	if resp.StatusCode != 200 {
+		return 0, errors.New(resp.Status)
+	}
+
+	data, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, err
+	}
+
+	result := struct {
+		ID int64 `json:"id"`
+	}{}
+	err = json.Unmarshal(data, &result)
+	return result.ID, err
+}
+
+// UpdateAnnotation updates an existing annotation with the Annotation it is passed
+func (c *Client) UpdateAnnotation(a *Annotation) (int64, error) {
```

Comment:
Is it possible to add a test for this function?

Reply:
@tonglil - Sure thing. I've added a test of `UpdateAnnotation` and I've also added a `PatchAnnotation` method. I have some ideas for improving the testing pattern (among a few other things), but I'm inclined to save such things for future PRs in the interest of keeping this PR relatively small and easy to review.

Reply:
Thank you for doing that